sql: split parser.Datum into Type and Datum

eisenstatdavid opened the pull request:
Currently we use Datum to represent both types and values, with naming
conventions that indicate the intended use. Having separate types is
clearer and will aid future changes to the type system.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/9912)

<!-- Reviewable:end -->

Commit:
sql: split parser.Datum into Type and Datum

Currently we use Datum to represent both types and values, with naming
conventions that indicate the intended use. Having separate types is
clearer and will aid future changes to the type system.

Remove DPlaceholder -- placeholders are variables, not values. To
compensate, Placeholder becomes a TypedExpr with type TPlaceholder.

Since we're touching adjacent code anyway, fix the Size computation for
DDecimal.

knz commented:
cc @dt

knz commented:
First review round done on `sql/parser`.
Can you briefly explain in the commit message what you did to Placeholder and DPlaceholder. Thanks.

---

Reviewed 26 of 66 files at r1.
Review status: 26 of 66 files reviewed at latest revision, 4 unresolved discussions.

---

*[sql/parser/expr.go, line 599 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9912#-KTtg-MJxs0kU_d4eGgP:-KTtg-ML28c4IyXOBXL5:b-67b6s2) ([raw file](https://github.com/cockroachdb/cockroach/blob/be8ea886cbe355901b22513a55cc0c8ed9420736/sql/parser/expr.go#L599)):*
> ```Go
> // ReturnType implements the TypedExpr interface.
> func (node *Tuple) ReturnType() Type {
> 	return TTuple{node.types}
> ```

:100: :star2: 

---

*[sql/parser/overload.go, line 85 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9912#-KTtiy4QXXe6hSyatvAI:-KTtiy4RclORkEi73Z4O:bviy2d1) ([raw file](https://github.com/cockroachdb/cockroach/blob/be8ea886cbe355901b22513a55cc0c8ed9420736/sql/parser/overload.go#L85)):*
> ```Go
> 	var s bytes.Buffer
> 	for i, typ := range a {
> 		if i != 0 {
> ```

use `i >  0` for consistency with the rest of the code base.

---

*[sql/parser/star_datum.go, line 57 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9912#-KTtgq4B7zOXyFNiJUQK:-KTtgq4DFt1u2cXbU_l4:b-8igz4q) ([raw file](https://github.com/cockroachdb/cockroach/blob/be8ea886cbe355901b22513a55cc0c8ed9420736/sql/parser/star_datum.go#L57)):*
> ```Go
> // Eval implements the TypedExpr interface.
> func (*StarDatum) Eval(ctx *EvalContext) (Datum, error) {
> 	return TypeInt.Exemplar().Eval(ctx)
> ```

Could you double check here whether this is the right approach. I think a simple `return DInt(0)` would be valid too, as would an error perhaps (we really should never evaluate a star).

---

*[sql/parser/type_check.go, line 35 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9912#-KTtiWLpLix-N1h_MGmH:-KTtiWLqwG6_GB8nowkT:b-kvfnia) ([raw file](https://github.com/cockroachdb/cockroach/blob/be8ea886cbe355901b22513a55cc0c8ed9420736/sql/parser/type_check.go#L35)):*
> ```Go
> 	Equal(other Type) bool
> 	// Exemplar returns a Datum whose type is the receiver.
> 	Exemplar() Datum
> ```

Where do we need this exactly?
If it's only for StarDatum perhaps we can remove it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9912#-:-KTtgLy-FfNcf9cZMtg_:b-2yp7zi)*
<!-- Sent from Reviewable.io -->

knz commented:
Done distsql, sqlbase. Looking good.

---

Reviewed 11 of 66 files at r1.
Review status: 37 of 66 files reviewed at latest revision, 5 unresolved discussions.

---

*[sql/testdata/aggregate, line 472 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9912#-KTtk-7ddlHeDq2dAQoO:-KTtk-7esBiRZiYTxUK1:b-2ihicb) ([raw file](https://github.com/cockroachdb/cockroach/blob/be8ea886cbe355901b22513a55cc0c8ed9420736/sql/testdata/aggregate#L472)):*
> ```
> SELECT AVG(c) FROM abc
> 
> query error unknown signature for AVG: AVG\(tuple{string, bool}\)
> ```

:purple_heart: :100:

---

*[sql/testdata/subquery, line 80 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9912#-KTtk73D1mAqhl6T7p99:-KTtk73D1mAqhl6T7p9A:b-2ihicb) ([raw file](https://github.com/cockroachdb/cockroach/blob/be8ea886cbe355901b22513a55cc0c8ed9420736/sql/testdata/subquery#L80)):*
> ```
> 7 8 9
> 
> statement error value type tuple{int, int, int} doesn't match type INT of column "a"
> ```

:purple_heart: :100:

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9912#-:-KTtkGSBhXYdFvc8qXnv:b-6femca)*
<!-- Sent from Reviewable.io -->

knz commented:
This was herculean work. Very nice. Thanks for tackling this, too.

---

Reviewed 29 of 66 files at r1.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[sql/create.go, line 1173 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9912#-KTtlMQGZE_5jqCB35kg:-KTtlMQIs_rR0d1lChFE:b-r9yc2j) ([raw file](https://github.com/cockroachdb/cockroach/blob/be8ea886cbe355901b22513a55cc0c8ed9420736/sql/create.go#L1173)):*
> ```Go
> 		}
> 		// Convert to a dummy datum of the correct type.
> 		return nil, false, col.Type.ToDatumType().Exemplar()
> ```

I'd use DNull here, despite what the comment says. Can you try that and see we're not breaking any test?

---

*[sql/pg_catalog.go, line 331 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9912#-KTtm_nwQu0uRpGgdds0:-KTtm_nymhrWEeKoPnSk:b168dx1) ([raw file](https://github.com/cockroachdb/cockroach/blob/be8ea886cbe355901b22513a55cc0c8ed9420736/sql/pg_catalog.go#L331)):*
> ```Go
> 
> func typLen(typ parser.Type) parser.Datum {
> 	if sz, variable := typ.Exemplar().Size(); !variable {
> ```

ditto, Size() should be method of Type.

---

*[sql/pgwire_test.go, line 269 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9912#-KTtmjbQF1lsVTxTP1d6:-KTtmjbRB3hUheYQtTt2:bt2xsam) ([raw file](https://github.com/cockroachdb/cockroach/blob/be8ea886cbe355901b22513a55cc0c8ed9420736/sql/pgwire_test.go#L269)):*
> ```Go
> 		"SELECT $1 > 0 AND NOT $1":                  "pq: incompatible NOT argument type: int",
> 		"CREATE TABLE $1 (id INT)":                  "pq: syntax error at or near \"1\"\nCREATE TABLE $1 (id INT)\n             ^\n",
> 		"UPDATE d.t SET s = i + $1":                 "pq: unsupported binary operator: <int> + <placeholder{1}> (desired <string>)",
> ```

:purple_heart: 

---

*[sql/row_container.go, line 76 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9912#-KTtkWNjUatzcgsR2A0d:-KTtkWNkZALtFQT3FCOV:bqcar35) ([raw file](https://github.com/cockroachdb/cockroach/blob/be8ea886cbe355901b22513a55cc0c8ed9420736/sql/row_container.go#L76)):*
> ```Go
> 
> 	for i := 0; i < nCols; i++ {
> 		sz, variable := h[i].Typ.Exemplar().Size()
> ```

I'd make Size() a method of the Type interface, much cleaner.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9912#-:-KTtn8ucY55sPOZ4KAC7:b-aj89ry)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: 64 of 66 files reviewed at latest revision, 7 unresolved discussions, some commit checks pending.

---

*[sql/parser/type_check.go, line 33 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9912#-KTtpAXCDkcJkpTQq6-r:-KTtpAXDXyk4betWH4d8:b-wint86) ([raw file](https://github.com/cockroachdb/cockroach/blob/590182b9e9723641ee07c68b2cb7b4e91467b697/sql/parser/type_check.go#L33)):*
> ```Go
> 	// Equal returns whether the receiver and the other type are the same. Use ==
> 	// for comparison against constant non-parametric types.
> 	Equal(other Type) bool
> ```

I don't grok this - when are the semantics of this method different from `==`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9912)*
<!-- Sent from Reviewable.io -->

knz commented:
On 12-10-16 12:58, Tamir Duberstein wrote:
> I don't grok this - when are the semantics of this method different from
> |==|?
> 

When there are tuples.


-- 
Raphael 'kena' Poss

tamird commented:
Ah, I see. Perhaps it would be better to make it impossible to compare types using `==` by embedding something inconvenient.

knz commented:
On 12-10-16 13:29, Tamir Duberstein wrote:
> Perhaps it would be better to make it impossible to compare
> types using |==| by embedding something inconvenient.

What type makes this possible in Go?

Or are you thinking about a linter check like we have for Lock()?


-- 
Raphael 'kena' Poss

tamird commented:
Concrete types which include slices or maps are incomparable.

On Wed, Oct 12, 2016 at 1:30 PM, kena <notifications@github.com> wrote:

> On 12-10-16 13:29, Tamir Duberstein wrote:
> > Perhaps it would be better to make it impossible to compare
> > types using |==| by embedding something inconvenient.
>
> What type makes this possible in Go?
>
> Or are you thinking about a linter check like we have for Lock()?
>
>
> --
> Raphael 'kena' Poss
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/9912#issuecomment-253281555>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABdsPCAe-X0BHItWDucoHQcGJkCElQxbks5qzRlOgaJpZM4KU7OE>
> .
>

petermattis commented:
You can embed a `[]struct{}`. Slices are not comparable.

On Wed, Oct 12, 2016 at 1:31 PM, kena <notifications@github.com> wrote:

> On 12-10-16 13:29, Tamir Duberstein wrote:
> > Perhaps it would be better to make it impossible to compare
> > types using |==| by embedding something inconvenient.
>
> What type makes this possible in Go?
>
> Or are you thinking about a linter check like we have for Lock()?
>
>
> --
> Raphael 'kena' Poss
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/9912#issuecomment-253281555>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AF6f95Opf56XEFIA_DiKC-UYifKhhduoks5qzRlUgaJpZM4KU7OE>
> .
>

eisenstatdavid commented:
The non-parametric types (i.e., all except Placeholder and Tuple) are all singletons. I don't want to ban == because it's less expensive than an interface method call and allows cleaner switch statements. We can't do away with .Equal() without interning parametric types, which we could do at the cost of some complication. I improved the documentation about when it's OK to use ==.

---

Review status: 62 of 66 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[sql/create.go, line 1173 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9912#-KTtlMQGZE_5jqCB35kg:-KTuB9iwqih-XdoyFr5H:bbpaa0g) ([raw file](https://github.com/cockroachdb/cockroach/blob/be8ea886cbe355901b22513a55cc0c8ed9420736/sql/create.go#L1173)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> I'd use DNull here, despite what the comment says. Can you try that and see we're not breaking any test?
</details>
Two tests break:
```
root@:26257> SELECT ABS(NULL);
pq: unknown signature for ABS: ABS(NULL)
```
I'll file an issue because this seems wrong.

---

*[sql/row_container.go, line 76 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9912#-KTtkWNjUatzcgsR2A0d:-KTuDB7xZ4_7hV_HOawj:b-p05xyq) ([raw file](https://github.com/cockroachdb/cockroach/blob/be8ea886cbe355901b22513a55cc0c8ed9420736/sql/row_container.go#L76)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> I'd make Size() a method of the Type interface, much cleaner.
</details>
Since Size() isn't constant for some types, I'm concerned about unintended consequences of this change. I'd rather defer it to a subsequent PR.

---

*[sql/parser/overload.go, line 85 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9912#-KTtiy4QXXe6hSyatvAI:-KTtkL8kpDfMUO_144vY:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/be8ea886cbe355901b22513a55cc0c8ed9420736/sql/parser/overload.go#L85)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> use `i >  0` for consistency with the rest of the code base.
</details>
Done.

---

*[sql/parser/star_datum.go, line 57 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9912#-KTtgq4B7zOXyFNiJUQK:-KTtl1TDlr8atUejmuap:b-4fe64s) ([raw file](https://github.com/cockroachdb/cockroach/blob/be8ea886cbe355901b22513a55cc0c8ed9420736/sql/parser/star_datum.go#L57)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> Could you double check here whether this is the right approach. I think a simple `return DInt(0)` would be valid too, as would an error perhaps (we really should never evaluate a star).
</details>
Mindlessly transliterated from the original TypeInt.Eval(). I replaced it with the return (the error crashes something; don't want to attempt that refactor in this PR).

---

*[sql/parser/type_check.go, line 35 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9912#-KTtiWLpLix-N1h_MGmH:-KTtlZt2h0YS8C8GBp3P:b5bcmra) ([raw file](https://github.com/cockroachdb/cockroach/blob/be8ea886cbe355901b22513a55cc0c8ed9420736/sql/parser/type_check.go#L35)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> Where do we need this exactly?
> If it's only for StarDatum perhaps we can remove it.
</details>
It gets used a couple places -- getting a size lower bound for a type is the main one, but there's another one in sql/create.go that I didn't take the time to understand.

---

*[sql/parser/type_check.go, line 33 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9912#-KTtpAXCDkcJkpTQq6-r:-KTuBfaf7aVrHWNqALwh:b-hmfq92) ([raw file](https://github.com/cockroachdb/cockroach/blob/590182b9e9723641ee07c68b2cb7b4e91467b697/sql/parser/type_check.go#L33)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> I don't grok this - when are the semantics of this method different from `==`?
</details>
Parameterized types (tuples and placeholders) instantiated with twice with the same parameters compare `Equal` but not `==`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9912#-:-KTuAWK2YKGIpXQnZYwz:bgydz87)*
<!-- Sent from Reviewable.io -->

knz commented:
Reviewed 2 of 2 files at r2, 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[sql/create.go, line 1173 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9912#-KTtlMQGZE_5jqCB35kg:-KTuI3D4q4iMwqspVlAE:b-tu9mjl) ([raw file](https://github.com/cockroachdb/cockroach/blob/be8ea886cbe355901b22513a55cc0c8ed9420736/sql/create.go#L1173)):*
<details><summary><i>Previously, eisenstatdavid (David Eisenstat) wrote…</i></summary>
> Two tests break:
> ```
> root@:26257> SELECT ABS(NULL);
> pq: unknown signature for ABS: ABS(NULL)
> ```
> I'll file an issue because this seems wrong.
</details>
Yes and put nathan in cc on that.

---

*[sql/row_container.go, line 76 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9912#-KTtkWNjUatzcgsR2A0d:-KTuIBLn0P62W9nx7Hl5:b1gkwzc) ([raw file](https://github.com/cockroachdb/cockroach/blob/be8ea886cbe355901b22513a55cc0c8ed9420736/sql/row_container.go#L76)):*
<details><summary><i>Previously, eisenstatdavid (David Eisenstat) wrote…</i></summary>
> Since Size() isn't constant for some types, I'm concerned about unintended consequences of this change. I'd rather defer it to a subsequent PR.
</details>
That's exactly what the interface for Size() indicates. Calling Size() on an "Exemplar" makes no sense to me...

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9912)*
<!-- Sent from Reviewable.io -->

eisenstatdavid commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[sql/row_container.go, line 76 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9912#-KTtkWNjUatzcgsR2A0d:-KTuIz01nIuLKaiCNoNc:b-paq3r) ([raw file](https://github.com/cockroachdb/cockroach/blob/be8ea886cbe355901b22513a55cc0c8ed9420736/sql/row_container.go#L76)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> That's exactly what the interface for Size() indicates. Calling Size() on an "Exemplar" makes no sense to me...
</details>
The problem is what to do with the .Size() calls directly on a Datum. The hypothetical d.ReturnType().Size() as compared to d.Size() necessarily changes the behavior in some cases where d is (e.g.) a decimal. I don't think .Exemplar().Size() is ridiculous; it's just a convenient proxy for "how big is an average datum of this type?"

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9912)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[sql/row_container.go, line 76 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9912#-KTtkWNjUatzcgsR2A0d:-KTuLbIyzv4Sk_H4ViZL:bj6gbic) ([raw file](https://github.com/cockroachdb/cockroach/blob/be8ea886cbe355901b22513a55cc0c8ed9420736/sql/row_container.go#L76)):*
<details><summary><i>Previously, eisenstatdavid (David Eisenstat) wrote…</i></summary>
> The problem is what to do with the .Size() calls directly on a Datum. The hypothetical d.ReturnType().Size() as compared to d.Size() necessarily changes the behavior in some cases where d is (e.g.) a decimal. I don't think .Exemplar().Size() is ridiculous; it's just a convenient proxy for "how big is an average datum of this type?"
</details>
It's obvious that the semantics of the Size() API must evolve:
- `Size()` on Type must return sz, bool
- `Size()` on Datum returns only a sz (no bool) and its corresponding implementation calls its Type.Size() if constant.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9912)*
<!-- Sent from Reviewable.io -->

eisenstatdavid commented:
Changed TTuple from a struct to a type alias. This seems to be a happier solution.

---

Review status: 56 of 66 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[sql/row_container.go, line 76 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9912#-KTtkWNjUatzcgsR2A0d:-KTueF0qmv2NtQ_N_vgE:bp8ublb) ([raw file](https://github.com/cockroachdb/cockroach/blob/be8ea886cbe355901b22513a55cc0c8ed9420736/sql/row_container.go#L76)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> It's obvious that the semantics of the Size() API must evolve:
> - `Size()` on Type must return sz, bool
> - `Size()` on Datum returns only a sz (no bool) and its corresponding implementation calls its Type.Size() if constant.
</details>
Done. I fixed a bug in the implementation of DDecimal.Size while I was at it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9912#-:-KTueG4m3_BKoOLBBC0s:byzhyqp)*
<!-- Sent from Reviewable.io -->

eisenstatdavid commented:
Review status: 56 of 66 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[sql/pg_catalog.go, line 331 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9912#-KTtm_nwQu0uRpGgdds0:-KTuezoISxB_jSpksdoI:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/be8ea886cbe355901b22513a55cc0c8ed9420736/sql/pg_catalog.go#L331)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> ditto, Size() should be method of Type.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9912)*
<!-- Sent from Reviewable.io -->

knz commented:
Reviewed 10 of 10 files at r4.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[sql/parser/datum.go, line 205 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9912#-KTugeBb9U1rfehh7Dh9:-KTugeBcT5hNQMSJufBV:b1x50si) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ddd649e01a969b78af019ee5b24b93aeb6ccd96/sql/parser/datum.go#L205)):*
> ```Go
> // Size implements the Datum interface.
> func (d *DBool) Size() uintptr {
> 	return unsafe.Sizeof(*d)
> ```

See my other comment. If Go does not partially evaluate Sizeof, these need to be precomputed.

---

*[sql/parser/type_check.go, line 108 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9912#-KTugSNnlPYsZwOhAXZg:-KTugSNnlPYsZwOhAXZh:blfayz2) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ddd649e01a969b78af019ee5b24b93aeb6ccd96/sql/parser/type_check.go#L108)):*
> ```Go
> func (tNull) Exemplar() Datum             { return DNull }
> func (tNull) FamilyEqual(other Type) bool { return other == TypeNull }
> func (tNull) Size() (uintptr, bool)       { return unsafe.Sizeof(dNull{}), fixedSize }
> ```

I am not sure that Go partially evaluates Sizeof. Does it? If not, pre-evaluate the sizes and return the constant here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9912)*
<!-- Sent from Reviewable.io -->

eisenstatdavid commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[sql/parser/type_check.go, line 108 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9912#-KTugSNnlPYsZwOhAXZg:-KTyVZfs7nhBbMIeE6yM:b-3vchnn) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ddd649e01a969b78af019ee5b24b93aeb6ccd96/sql/parser/type_check.go#L108)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> I am not sure that Go partially evaluates Sizeof. Does it? If not, pre-evaluate the sizes and return the constant here.
</details>
The output of `go tool compile -S` on the following program suggests that Go is able to do that optimization, since the two `println` statements compile identically:
```
package main

import "unsafe"

type DInt int64

func main() {
	println(uintptr(8))
	println(unsafe.Sizeof(DInt(0)))
}
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9912)*
<!-- Sent from Reviewable.io -->

eisenstatdavid commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[sql/parser/datum.go, line 205 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9912#-KTugeBb9U1rfehh7Dh9:-KTyVy24-vxzQzT5LPxs:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ddd649e01a969b78af019ee5b24b93aeb6ccd96/sql/parser/datum.go#L205)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> See my other comment. If Go does not partially evaluate Sizeof, these need to be precomputed.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9912)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[sql/parser/type_check.go, line 108 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9912#-KTugSNnlPYsZwOhAXZg:-KTybNf3fM2jlkPdh4NC:b-ucfknh) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ddd649e01a969b78af019ee5b24b93aeb6ccd96/sql/parser/type_check.go#L108)):*
<details><summary><i>Previously, eisenstatdavid (David Eisenstat) wrote…</i></summary>
> The output of `go tool compile -S` on the following program suggests that Go is able to do that optimization, since the two `println` statements compile identically:
> ```
> package main
> 
> import "unsafe"
> 
> type DInt int64
> 
> func main() {
> 	println(uintptr(8))
> 	println(unsafe.Sizeof(DInt(0)))
> }
> ```
</details>
Thanks!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9912)*
<!-- Sent from Reviewable.io -->

knz commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

Let's see what Nathan thinks of this too.

---

Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9912#-:-KTybUBPhupTHWVd7Bqf:b2e8osr)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> This is great! @knz and I have been dreaming about a split like this for months. Thanks for all the work here!

---

Reviewed 54 of 66 files at r1, 2 of 2 files at r2, 10 of 10 files at r4.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[sql/parser/expr.go, line 60 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9912#-KTypFbDE6mlmjr-lrJU:-KTypFbDE6mlmjr-lrJV:b-8kl4wl) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ddd649e01a969b78af019ee5b24b93aeb6ccd96/sql/parser/expr.go#L60)):*
> ```Go
> 	// ReturnType provides the type of the TypedExpr, which is the type of Datum that
> 	// the TypedExpr will return when evaluated.
> 	ReturnType() Type
> ```

After making this change, how do you feel about the name `ReturnType`? I could see us changing it to something like `ResolvedType` to better express the method's meaning of returning the type our type checker inferred.

---

*[sql/parser/type_check.go, line 29 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9912#-KTykyObsq5Z40qbT6yV:-KTykyObsq5Z40qbT6yW:b-pe6yoz) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ddd649e01a969b78af019ee5b24b93aeb6ccd96/sql/parser/type_check.go#L29)):*
> ```Go
> )
> 
> // Type represents a SQL type.
> ```

Consider moving all of this to a `types.go` file

---

*[sql/parser/type_check.go, line 38 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9912#-KTyoNQzAypiqtQVexBn:-KTyoNQzAypiqtQVexBo:bd051lp) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ddd649e01a969b78af019ee5b24b93aeb6ccd96/sql/parser/type_check.go#L38)):*
> ```Go
> 	// Exemplar returns a Datum whose type is the receiver.
> 	//
> 	// TODO(eisen): remove this method.
> ```

How difficult do you foresee this being? Is it just a matter of threading `parser.Type` in place of `parser.Datum` in a few places in the `sql` package?

---

*[sql/parser/type_check.go, line 42 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9912#-KTykEZJ2a8KmNm68yYY:-KTykEZJ2a8KmNm68yYZ:bewbhe) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ddd649e01a969b78af019ee5b24b93aeb6ccd96/sql/parser/type_check.go#L42)):*
> ```Go
> 	// FamilyEqual returns whether the receiver and the other type have the same
> 	// constructor.
> 	FamilyEqual(other Type) bool
> ```

nit here and above: parameter names (`other`) are not needed in interface definitions.

---

*[sql/parser/type_check.go, line 252 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9912#-KTylAVAgZa3Qns5D6hp:-KTylAVAgZa3Qns5D6hq:bq5rudu) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ddd649e01a969b78af019ee5b24b93aeb6ccd96/sql/parser/type_check.go#L252)):*
> ```Go
> 
> // TPlaceholder is the type of a placeholder.
> type TPlaceholder struct{ Name string }
> ```

nit: `Name` should be on a separate line

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9912#-:-KTyllKDEbvGrfMVMonz:b-9wnivo)*
<!-- Sent from Reviewable.io -->

knz commented:
This needs to be rebased but assuming CI validates after rebase it is good to merge.

eisenstatdavid commented:
PTAL -- did the refactoring to remove Exemplar().

---

Review status: 30 of 67 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[sql/parser/expr.go, line 60 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9912#-KTypFbDE6mlmjr-lrJU:-KUIHX1WIIopwC4pqTl8:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ddd649e01a969b78af019ee5b24b93aeb6ccd96/sql/parser/expr.go#L60)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> After making this change, how do you feel about the name `ReturnType`? I could see us changing it to something like `ResolvedType` to better express the method's meaning of returning the type our type checker inferred.
</details>
Done.

---

*[sql/parser/type_check.go, line 29 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9912#-KTykyObsq5Z40qbT6yV:-KUIkS2h1caPXZHaPoj3:b-tu2vrw) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ddd649e01a969b78af019ee5b24b93aeb6ccd96/sql/parser/type_check.go#L29)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> Consider moving all of this to a `types.go` file
</details>
Will do this once I get around to rebasing -- don't want to screw up the review history prematurely.

---

*[sql/parser/type_check.go, line 38 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9912#-KTyoNQzAypiqtQVexBn:-KUIfcpbKiqpGDMTnqcK:b-jp9fq0) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ddd649e01a969b78af019ee5b24b93aeb6ccd96/sql/parser/type_check.go#L38)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> How difficult do you foresee this being? Is it just a matter of threading `parser.Type` in place of `parser.Datum` in a few places in the `sql` package?
</details>
Not too bad -- I added a new expression node for one call site and refactored the other.

---

*[sql/parser/type_check.go, line 42 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9912#-KTykEZJ2a8KmNm68yYY:-KUII6neTsWhJ4wVZOVj:b8dkuud) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ddd649e01a969b78af019ee5b24b93aeb6ccd96/sql/parser/type_check.go#L42)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> nit here and above: parameter names (`other`) are not needed in interface definitions.
</details>
Not needed, but useful when creating new interface definitions for consistency.

---

*[sql/parser/type_check.go, line 252 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9912#-KTylAVAgZa3Qns5D6hp:-KUIfaM8cafFKFH24mad:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ddd649e01a969b78af019ee5b24b93aeb6ccd96/sql/parser/type_check.go#L252)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> nit: `Name` should be on a separate line
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9912#-:-KUIkdjx-chTk8-6QKY3:b-ey7lzq)*
<!-- Sent from Reviewable.io -->

knz commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> again.

---

Reviewed 37 of 37 files at r5.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[sql/create.go, line 1127 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/9912#-KUJ0Xz1yHsNIwN9uxcg:-KUJ0Xz1yHsNIwN9uxch:b-gm8ssz) ([raw file](https://github.com/cockroachdb/cockroach/blob/a48337d637652121181e91569f53c31b56465cfb/sql/create.go#L1127)):*
> ```Go
> }
> 
> type dummyColumnItem struct {
> ```

It would be useful to add an explanatory comment here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9912#-:-KUJ1MZy1KN1uVVPW5ut:butqosd)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
Review status: 0 of 68 files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---

*[pkg/sql/sqlbase/table.go, line 1136 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/9912#-Kak8H8jxynV9fnLnGzU:-Kak8H8jxynV9fnLnGzV:b-3mzdlz) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3e6319c27ebf07be4c048baac2c6bd1a785d1b3/pkg/sql/sqlbase/table.go#L1136)):*
> ```Go
> 				p.Name, col.Name, err)
> 		}
> 	} else if !(typ.Equal(set) || (set == parser.TypeBytes && typ == parser.TypeString)) {
> ```

@eisenstatdavid could you explain the `set == parser.TypeBytes && typ == parser.TypeString` part to me? I just stumbled upon the code and can't see why it's needed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9912)*
<!-- Sent from Reviewable.io -->

eisenstatdavid commented:
*[pkg/sql/sqlbase/table.go, line 1136 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/9912#-Kak8H8jxynV9fnLnGzU:-KamAegDWFmfBhJNM7-N:b9omlpn) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3e6319c27ebf07be4c048baac2c6bd1a785d1b3/pkg/sql/sqlbase/table.go#L1136)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

@eisenstatdavid could you explain the `set == parser.TypeBytes && typ == parser.TypeString` part to me? I just stumbled upon the code and can't see why it's needed.
</blockquote></details>

I guess we want to accept strings in bytes columns? I moved the code from the ColumnType_BYTES case without thinking about it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9912)*
<!-- Sent from Reviewable.io -->
